readme.md
# GLOpenAPI: NASA GeneLab Open API

__(Formerly known as GeneFab3)__

Provides cross-dataset access to multi-omics annotation and data from the
[GeneLab data repository](https://genelab-data.ndc.nasa.gov/genelab/projects);
is currently deployed at https://visualization.genelab.nasa.gov/GLOpenAPI/ and
powers the [GeneLab data visualization dashboard](https://visualization.genelab.nasa.gov/data/).

Note: if you intend to use it for downstream applications and analyses, there is
no need to install anything yourself; skip to the [Usage](#usage) section for
more information.


## Installation

It is possible, however, to deploy GLOpenAPI yourself as well.


### Requirements

* Python 3 (developed and tested on 3.6.10)
* MongoDB (developed and tested on 4.0.2-2-gddce0aa9fc)
* SQLite3 (developed and tested on 3.32.3 2020-06-18 14:00:33)
* pymongo
* flask
* flask-compress
* isatools
* numpy
* pandas
* natsort
* filelock


### How the GLOpenAPI backend operates

In general, GLOpenAPI:
* Maintains one daemon thread which continuously caches and updates metadata for
  all GeneLab datasets, and stores it in a MongoDB database
  (`db_name="genefab3"` by default):
  * investigation, samples, and assays annotation from datasets' ISA ZIP files,
  * and all publicly available file names and URLs associated with each dataset.
* Caches data on-demand (i.e., the first time it is requested by a user and then
  any time the source file is updated in the GeneLab repository) in SQLite3
  databases (which, of course, should be writable by GLOpenAPI); by default:
  * `.genefab3.sqlite3/blobs.db` for binary blobs and
  * `.genefab3.sqlite3/tables.db` for tabular data.
* Caches tabular data *representations* on-demand (for example, JSON-formatted
  merged unnormalized RNA-Seq counts, or a CSV-formatted differential analysis
  table subset by `Amean > 10`, etc) &ndash; again, the first time a specific
  query is requested, and then any time the source file(s) are updated;
  * `.genefab3.sqlite3/response-cache.db` by default.

Please refer to the comments in the `app.py` file for configuration options and
for the birds-eye view of the backend.


#### mod_wsgi primer

The installation process is typical for a Flask app, e.g.
with [mod_wsgi](https://flask.palletsprojects.com/en/2.0.x/deploying/mod_wsgi/).
Note that the name of the singleton instance of the app is called `flask_app`,
therefore in the `.wsgi` file you will need something similar to the following
line of code: `from app import flask_app as app`.


## Usage

Please refer to the description of the structure and the query syntax over at
the [API landing page](https://visualization.genelab.nasa.gov/GLOpenAPI/); in
JavaScript-capable browsers it also provides an interactive URL builder.

Basic (and very naive) examples of using the API programmatically (e.g., in
Python) are available in this repository under
[./examples](https://github.com/LankyCyril/GLOpenAPI/tree/master/examples);
for instance,
[heatmap.ipynb](https://github.com/LankyCyril/GLOpenAPI/tree/master/examples/heatmap.ipynb).


changelog.md
Version 4.1.2-alpha0 (2023-10-03)

- Fix:
    - New logic that matches Assay entries to Study entries: if unable to guess
      the matching sample name, check if there's only a single Study file in the
      ISA Zip (actually, currently this is always the case), and use that
      information instead.
      Old logic (guessing the best sample name match) is retained to be future-
      proof in case ISA Zips with multiple Study entries are added to the
      repository later.


Version 4.1.1-alpha0 (2023-08-14)

- Hotfix:
    - For datasets with RSEM and STAR unnormalized counts, pick RSEM by default


Version 4.1.0-alpha1 (2023-08-02)

- Internal:
    - Use `pyximport` machinery instead of calling `cythonize` to build
      cython submodules on the fly


Version 4.1.0-alpha0 (2023-08-01)

- Performance improvements:
    - Use Cython to speed up normalization of MongoDB documents with known
      schema:
        - improves the performance of /metadata-counts/ endpoints
        - improves the performance of /samples/ endpoints
    - Implement `FuncTee()` in lieu of `itertools.tee()` to significantly reduce
      RAM usage at the expense of re-running the underlying generating function
- Fixes:
    - Reinstate check_isa_elements=True and fix the ISA parser to split
      file lists correctly; fixes wrong extraneous files being assigned to
      samples (e.g. FASTQs of Sample A assigned both to Sample A and Sample B
      within the same study)
    - Enumerate all bottom-level leaves in /metadata-counts/ branches
- Internal:
    - The debug launcher accepts parameters to control the activity of the
      cacher thread and of the response cache separately
    - More logging in development/debug mode
- Temporarily disabled:
    - Interactive widgets on the landing page that rely on outdated and
      performance-intensive database transactions; relevant syntax examples are
      retained on the landing page
    - Corresponding background functionality that used to update auxiliary
      MongoDB collections and waste CPU cycles
    NOTE: These are disabled pending backend redesign and performance
      improvements.
    NOTE: This is not a breaking change as all the API functionality is intact;
      however, the landing page loses some level of interactivity with little to
      no detriment to the presentation of the capabilities of the query syntax.


Version 4.0.11-alpha0 (2023-07-20)

- Fixes:
    - Different MongoDB query validation rules for user-passed and internally-
      generated queries
    - Reduce the number of MongoDB requests (sample name lookups) when caching
      data tables


Version 4.0.10-alpha0 (2023-07-05)

- Internal:
    - Pre-cache the response to a smaller /metadata-cache/ request (the one used
      in the visualization app):
        investigation.study%20assays.study%20assay%20technology%20type&
        investigation.study.comment.mission%20name&
        investigation.study.comment.project%20type&
        investigation.study.study%20title&
        study.characteristics.material%20type&
        study.characteristics.organism&
        study.factor%20value


Version 4.0.9-alpha0 (2023-06-27)

- Fixes:
    - By default, do not check filename lists against the filenames present in
      the ISA record
    - Retain ability to optionally check this via parameter to
      Sample._INPLACE_extend_with_dataset_files(self, check_isa_elements: bool)
- Internal:
    - Drop the MongoDB database if it was created with an app version older than
      specified by parameter `min_app_version`


Version 4.0.8-alpha3 (2023-06-24)

- Hotfixes:
    - Actually pass flags=IGNORECASE to the regexes that are generated from
      plaintext values


Version 4.0.8-alpha2 (2023-05-24)

- Hotfixes:
    - Match both "rna_seq" and "rna-seq" in filenames
    - In debug mode, serialize functions when logging (print function name only)


Version 4.0.8-alpha1 (2023-05-22)

- Hotfix: in "all_files=True", look for "date_updated" (rather than
  "date_modified")


Version 4.0.8-alpha0 (2023-05-05)

- Internal:
    - Drop response_cache.db if it was created with an app version older than
      specified by parameter `min_app_version`


Version 4.0.7-alpha0 (2023-04-24)

- Fixes (internal):
    - Implicit metadata value regex matches start-to-END
- Internal:
    - Exhaustive debugging messages when enumerating files by accession


Version 4.0.6-alpha0 (2023-03-30)

- Features:
    - Case-insensitive search for metadata values (e.g.,
      "&study.factor value.spaceflight=ground Control|vivarIuM ConTrOL")
    - In debug/staging mode, the GENELAB_ROOT environment variable can be set
      to a different core NASA API endpoint (e.g. internal staging)
    - Support OSD accession IDs, simply by way of not assuming that they should
      start with anything particular. Any /(\S+)-(\d+)/ identifier is valid.
- Features, internal, staged for testing:
    - Make use of the new (to be rolled out) /files/ endpoint structure
      (request "all_files=True"; parse "date_created" and "date_modified")


Version 4.0.5-alpha0 (2023-02-06)

- Reverted:
    - Revert to using the /filelistings/ endpoint
- Fixes:
    - Use sha1 instead of md5 for FIPS compliance
    - Update logo in header on landing page, use OSDR style (logo only on left)


Version 4.0.4-alpha0 (2022-09-13)

- Hotfix: serve `root.js` with correct mimetype (application/javascript)


Version 4.0.3-alpha0 (2022-09-06)

- Fixes (internal):
    - Use new NASA API endpoints
    - Enforce new Content Security Policy rules;
      factor out inline JavaScript and CSS
- Regression:
    - Temporarily disable '&format=browser'; usage of SlickGrid currently
      violates the Content Security Policy


Version 4.0.2-alpha0 (2022-07-26)

- Improvements:
    - /metadata-counts/ JSONs now list the number of accessions, assays, and
      samples at each level; each next level is available under the "children"
      key. For example:

genefab3/config.py
## Flask config

DEBUG_MARKERS = {"development", "staging", "stage", "debug", "debugging"}

COMPRESSIBLE_MIMETYPES = [
    "text/plain", "text/html", "text/css", "text/xml",
    "application/json", "application/javascript"
]

DEFAULT_FORMATS = {
    "/assays/": "tsv", "/samples/": "tsv", "/data/": "tsv", "/files/": "tsv",
}


## Databases

MONGO_CLIENT_PARAMETERS = dict(
    serverSelectionTimeoutMS=2000,
)
MONGO_DB_NAME = "genefab3"
MONGO_DB_LOCALE = "en_US"

MAX_JSON_AGE = 10800 # 3 hours (in seconds)
CACHER_THREAD_CHECK_INTERVAL = 1800 # 30 minutes (in seconds)
CACHER_THREAD_RECHECK_DELAY = 300 # 5 minutes (in seconds)
METADATA_INDEX_WAIT_DELAY = 60 # 1 minute (in seconds)
METADATA_INDEX_WAIT_STEP = 5

from types import SimpleNamespace
COLLECTION_NAMES = SimpleNamespace(
    STATUS="status",
    LOG="log",
    JSON_CACHE="json_cache",
    DATASET_TIMESTAMPS="dataset_timestamps",
    METADATA="metadata",
    METADATA_VALUE_LOOKUP="metadata_value_lookup",
    FILE_DESCRIPTORS="file_descriptors",
)

SQLITE_DB = "./.genefab3.sqlite/data.db"

RESPONSE_CACHE = "./.genefab3.sqlite/response-cache.db"
USE_RESPONSE_CACHE = True
ZLIB_COMPRESS_RESPONSE_CACHE = True
RESPONSE_CACHE_MAX_SIZE = 24*1024*1024*1024 # 24 GiB

from werkzeug.datastructures import ImmutableDict
RESPONSE_CACHE_SCHEMAS = ImmutableDict({
    "response_cache": """(
        'context_identity' TEXT, 'api_path' TEXT, 'timestamp' INTEGER,
        'response' BLOB, 'nbytes' INTEGER, 'mimetype' TEXT
    )""",
    "accessions_used": """(
        'context_identity' TEXT, 'accession' TEXT
    )""",
})


## External APIs

GENELAB_ROOT = "https://genelab-data.ndc.nasa.gov"
COLD_API_ROOT = "https://genelab-data.ndc.nasa.gov/genelab"
COLD_SEARCH_MASK = COLD_API_ROOT + "/data/search/?term=GLDS&type=cgene&size={}"
COLD_GLDS_MASK = COLD_API_ROOT + "/data/study/data/{}/"
COLD_FILEURLS_MASK = COLD_API_ROOT + "/data/glds/files/{}"
COLD_FILEDATES_MASK = COLD_API_ROOT + "/data/study/filelistings/{}"

TIMESTAMP_FMT = "%a %b %d %H:%M:%S %Z %Y"


## GeneFab API parser parameters

from operator import eq, ne, gt, getitem, contains, length_hint
not_in = lambda v, s: v not in s
listlen = lambda d, k: len(d.getlist(k))
leaf_count = lambda d, h: sum(length_hint(v, h) for v in d.values())

DISALLOWED_CONTEXTS = [
    dict(_="at least one dataset or annotation category must be specified",
        view=(eq, "/status/", eq, False),
        projection=(length_hint, 0, eq, 0), # no projection
        accessions_and_assays=(length_hint, 0, eq, 0), # no datasets
    ),
    dict(_="metadata queries are not valid for /status/",
        view=(eq, "/status/", eq, True), query=(leaf_count, 0, gt, 0),
    ),
    dict(_="'format=cls' is only valid for /samples/",
        view=(eq, "/samples/", eq, False), kwargs=(getitem, "format", eq, "cls"),
    ),
    dict(_="/data/ requires a 'datatype=' argument",
        view=(eq, "/data/", eq, True), kwargs=(contains, "datatype", eq, False),
    ),
    dict(_="'format=gct' is only valid for /data/",
        view=(eq, "/data/", eq, False), kwargs=(getitem, "format", eq, "gct"),
    ),
    dict(_="'format=gct' is not valid for the requested datatype",
        kwargs=[
            (getitem, "format", eq, "gct"),
            (getitem, "datatype", not_in, {"unnormalized counts"}),
        ],
    ),
    dict(_="/file/ only accepts 'format=raw'",
        view=(eq, "/file/", eq, True), kwargs=(getitem, "format", ne, "raw"),
    ),
    dict(_="/file/ requires at most one 'filename=' argument",
        view=(eq, "/file/", eq, True), kwargs=(listlen, "filename", gt, 1),
    ),
    dict(_="/file/ requires a single dataset in the 'from=' argument",
        view=(eq, "/file/", eq, True),
        accessions_and_assays=(length_hint, 0, ne, 1), # no. of datasets != 1
    ),
    dict(_="/file/ metadata categories are only valid for lookups in assays",
        view=(eq, "/file/", eq, True),
        accessions_and_assays=(leaf_count, 0, eq, 0), # no. of assays == 0
        projection=(length_hint, 0, gt, 0), # projection present
    ),
    dict(_="/file/ accepts at most one metadata category for lookups in assays",
        view=(eq, "/file/", eq, True),
        accessions_and_assays=(leaf_count, 0, eq, 1), # no. of assays == 1
        projection=(length_hint, 0, gt, 1), # many fields to look in
    ),
    dict(_="/file/ requires at most one assay in the 'from=' argument",
        view=(eq, "/file/", eq, True),
        accessions_and_assays=(leaf_count, 0, gt, 1), # no. of assays > 1
    ),
]


## (Meta)data parameters

ISA_ZIP_REGEX = r'.*_metadata_.*[_-]ISA\.zip$'
ANNOTATION_CATEGORIES = {"factor value", "parameter value", "characteristics"}
UNITS_FORMAT = "{value} {{{unit}}}"
ISA_TECH_TYPE_LOCATOR = "investigation.study assays.study assay technology type"

from collections import namedtuple
Locator = namedtuple("Locator", ["keys", "regex", "nargs"])

TECHNOLOGY_FILE_LOCATORS = {
    "rna sequencing (rna-seq)": {
        "unnormalized counts": Locator(
            keys=(
                "assay.parameter value.raw counts data file..",
                "assay.characteristics.raw counts data file..",
            ),
            regex=r'rna_seq_Unnormalized_Counts\.csv$',
            nargs=float("inf"),
        ),
        "differential expression": Locator(
            keys=(),
            regex=r'^GLDS-[0-9]+_(array|rna_seq)(_all-samples)?_differential_expression\.csv$',
            nargs=1,
        ),
        "visualization table": Locator(
            keys=(),
            regex=r'^GLDS-[0-9]+_(array|rna_seq)(_all-samples)?_visualization_output_table\.csv$',
            nargs=1,
        ),
    }
}

from collections import defaultdict
ROW_TYPES = defaultdict(lambda: "entry", {
    "unnormalized counts": "entry",
})

RAW_FILE_REGEX = r'file|plot'
PCA_CSV_REGEX = r'^GLDS-[0-9]+_(array|rna_seq)(_all-samples)?_visualization_PCA_table.csv$'
